File: 713_subarray_product_less_than_K.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        int i{0}, count{0}, product{1};
        for (int j = 0; j < nums.size(); j++){
            product *= nums[j];
            while (i <= j and product >= k)
                product /= nums[i++];
            count += (j - i + 1);
        }
        return count;
    }
};

File: 713_subarray_product_less_than_K.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        i, prod, count = 0, 1, 0
        for j, n in enumerate(nums):
            prod *= n
            while i <= j and prod >= k:
                prod /= nums[i]
                i += 1
            count += j - i + 1
        return count

